From 1c21fd613ae5c4c4a8e68f383f3bb1d92195b6fe Mon Sep 17 00:00:00 2001
From: Neeraj Singh <neerajdotname@gmail.com>
Date: Mon, 15 Nov 2010 17:52:39 -0500
Subject: [PATCH] If a user wants json output then try best to
 render json output. In such cases
 prefer kind_of(String) over respond_to?(to_str)

[#5841 state:resolved]
---
 .../lib/action_controller/metal/renderers.rb       |    2 +-
 actionpack/test/controller/mime_responds_test.rb   |   18 ++++++++++++++++++
 actionpack/test/lib/controller/fake_models.rb      |    6 ++++++
 3 files changed, 25 insertions(+), 1 deletions(-)

diff --git a/actionpack/lib/action_controller/metal/renderers.rb b/actionpack/lib/action_controller/metal/renderers.rb
index f9b226b..dc39e03 100644
--- a/actionpack/lib/action_controller/metal/renderers.rb
+++ b/actionpack/lib/action_controller/metal/renderers.rb
@@ -71,7 +71,7 @@ module ActionController
     end
 
     add :json do |json, options|
-      json = json.to_json(options) unless json.respond_to?(:to_str)
+      json = json.to_json(options) unless json.kind_of?(String)
       json = "#{options[:callback]}(#{json})" unless options[:callback].blank?
       self.content_type ||= Mime::JSON
       self.response_body  = json
diff --git a/actionpack/test/controller/mime_responds_test.rb b/actionpack/test/controller/mime_responds_test.rb
index b6ce9f7..a39f003 100644
--- a/actionpack/test/controller/mime_responds_test.rb
+++ b/actionpack/test/controller/mime_responds_test.rb
@@ -556,6 +556,17 @@ class InheritedRespondWithController < RespondWithController
   end
 end
 
+class RenderJsonRespondWithController < RespondWithController
+  clear_respond_to
+  respond_to :json
+
+  def index
+    respond_with(resource) do |format|
+      format.json { render :json => RenderJsonTestException.new('boom') }
+    end
+  end
+end
+
 class EmptyRespondWithController < ActionController::Base
   def index
     respond_with(Customer.new("david", 13))
@@ -869,6 +880,13 @@ class RespondWithControllerTest < ActionController::TestCase
     assert_equal "JSON", @response.body
   end
 
+  def test_render_json_object_responds_to_str_still_produce_json
+    @controller = RenderJsonRespondWithController.new
+    @request.accept = "application/json"
+    get :index, :format => :json
+    assert_equal %Q{{"message":"boom","error":"RenderJsonTestException"}}, @response.body
+  end
+
   def test_no_double_render_is_raised
     @request.accept = "text/html"
     assert_raise ActionView::MissingTemplate do
diff --git a/actionpack/test/lib/controller/fake_models.rb b/actionpack/test/lib/controller/fake_models.rb
index ae0c381..bd18cdc 100644
--- a/actionpack/test/lib/controller/fake_models.rb
+++ b/actionpack/test/lib/controller/fake_models.rb
@@ -194,3 +194,9 @@ class ArelLike
     a.each { |i| yield i }
   end
 end
+
+class RenderJsonTestException < Exception
+  def to_json(options = nil)
+    return { :error => self.class.name, :message => self.to_str }.to_json
+  end
+end
-- 
1.7.2

